pkg/dns/config/sync.go

```diff
@@ -0,0 +1,200 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package config
+
+import (
+	"k8s.io/client-go/pkg/util/wait"
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/client/cache"
+	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
+	fed "k8s.io/kubernetes/pkg/dns/federation"
+	"k8s.io/kubernetes/pkg/fields"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"time"
+
+	"github.com/golang/glog"
+)
+
+// Sync manages synchronization of the config map.
+type Sync interface {
+	// Once does a blocking synchronization of the config map. If the
+	// ConfigMap fails to validate, this method will return nil, err.
+	Once() (*Config, error)
+
+	// Start a periodic synchronization of the configuration map. When a
+	// successful configuration map update is detected, the
+	// configuration will be sent to the channel.
+	//
+	// It is an error to call this more than once.
+	Periodic() <-chan *Config
+}
+
+// NewSync for ConfigMap from namespace `ns` and `name`.
+func NewSync(client clientset.Interface, ns string, name string) Sync {
+	sync := &kubeSync{
+		ns:      ns,
+		name:    name,
+		client:  client,
+		channel: make(chan *Config),
+	}
+
+	listWatch := &cache.ListWatch{
```

Comment:
I wrote the code both ways and the complexity is not much different. Using the official Kubernetes watcher API has build in support for backoffs due to error etc.